### DISCUSSION TITLE
Sensor vs static object

### DISCUSSION THREAD

**Discussion by DocAce** [jrouwe/JoltPhysics#431] [Q&A] [answered]
In our game area damage is using a sensor, and it needs to also work against non-moving things like space stations. I noticed that there is a hardcoded check that prevents any collisions between sensors and static objects. Could that potentially become a setting that I can change?
As a workaround I can change the stations to be kinematic, but as they're made from several collision bodies that means a bunch of unnecessary collision contacts all the time.

**Comment by jrouwe** [comment, marked as answer]
Hello, it would probably be possible to support this, but it would not be the most efficient way to do things. Basically what would happen if you create a sensor that collides with static is that the physics simulation would do:

* parallel for each sensor
  * do a narrow phase query using the sensors shape and transform to find all colliding geometry
    * keep track of added/removed contact points in the contact cache
       * send callback notifications for these contact points

The contact cache is a big bottleneck in the physics simulation so adding all these extra contact points is not going to help performance, and you don't really need to use it for damage areas. I think you could implement it more efficiently yourself by doing:

* parallel for each damage area
   * do a narrow phase query using the shape and transform of the damage area, passing in broadphase/layer/body/shape filters to filter out irrelevant stuff as quickly as possible 
      * for each hit apply damage (possibly you need to keep track of multiple hits on the same body to avoid doing damage twice, possibly doing an extra ray cast towards the hit from the center of the damage area to double check that the object can actually be hit by the damage area)

**Reply by DocAce** [reply]
I see your point. I'm usually wary of doing too much stuff outside of the simulation step and losing the benefits of internal optimization structures. But if the contact cache is a significant bottleneck then doing this for area damage specifically sounds like the right approach. Thanks!

**Comment by jankrassnigg** [comment]
Why don't you use a regular shape query? I assume your area damage is something that stays active for a long time, so a sensor makes more sense? You could combine the two, and do a shape query to only find the static objects. You only need to repeat that when your damage area moves (far enough).

**Reply by DocAce** [reply]
It actually doesn't live long at all! I'll be trying out this approach, thanks :)

**Reply by jankrassnigg** [reply]
If it's literally a one-off thing like an explosion, shape queries are much better. You get the result right away (no contact add/remove handling etc) and the physics engine can't optimize anything anyway. And you can configure all the filters such that you can find all static / kinematic / dynamic bodies that you like.

**Reply by DocAce** [reply]
This is 99% about explosions, yes, but they can still potentially have a longer duration, especially if they're really big. And since our game gets modded a lot these should remain supported. For now I'll go with the safe and easy route of doing a query right away, but keeping the sensor up for the defined duration. I'll mark this as a potential optimization for later, though, skipping the physics body if the duration is zero.
